src/cca/scripts/uproxy_server.ts

```diff
@@ -56,25 +36,146 @@ export class UproxyServer implements Server {
+const CONTACTS_STORAGE_KEY = 'cloud-server-invites';
+
 export class UproxyServerRepository implements ServerRepository {
+  // core must already be logged into social networks.
   constructor(private core: CoreConnector, private vpnDevice: VpnDevice) {
-    this.core.login({
-      network: 'Cloud',
-      loginType: uproxy_core_api.LoginType.INITIAL,
+    if (!isStorageAvailable) {
+      console.warn('localStorage unavailable - contacts will not be saved');
+      return;
+    }
+  }
+
+  public onServerCallback :OnServerCallback;
+  public onServer(callback:OnServerCallback) {
+    this.onServerCallback = callback;
+    return this;
+  }
+
+  public restore() {
+    if (!this.onServerCallback) {
+      throw new Error('must call onServer first');
+    }
+
+    for (const savedContact of this.loadContacts().contacts) {
+      this.informCoreOfServer(savedContact.invite).then((server) => {
+        try {
+          this.onServerCallback(server);
+        } catch (e) {
+          console.warn('onServer callback threw', e);
+        }
+      }, (e) => {
+        console.warn('could not inform core of saved server', e);
+      });
+    }
+  }
+
+  // returns saved contacts, which will be empty if
+  // localStorage is unavailable or no contacts are found.
+  // throws if there is an error reading from localStorage or
+  // the stored contacts could not be parsed.
+  private loadContacts(): cloud_social_provider.SavedContacts {
+    if (!isStorageAvailable) {
+      return {
+        contacts: []
+      };
+    }
+
+    const savedContactsJson = localStorage.getItem(CONTACTS_STORAGE_KEY);
+    return savedContactsJson ? <cloud_social_provider.SavedContacts>JSON.parse(
+        savedContactsJson) : {
+      contacts: []
+    };
+  }
+
+  // merges a contact with those already saved to storage.
+  // returns true if the contact was not previously in storage.
+  // contacts are saved as a SavedContacts serialised as JSON. 
+  private saveContact(newInvite:cloud_social_provider.Invite) {
+    if (!isStorageAvailable) {
+      return true;
+    }
+
+    const savedContacts = this.loadContacts();
+
+    // merge or append, as necessary.
+    const dupeIndex = _.findIndex(savedContacts.contacts, (savedContact) => {
+      return savedContact.invite && savedContact.invite.host === newInvite.host;
     });
+
+    const newSavedContact: cloud_social_provider.SavedContact = {
+      invite: newInvite
+    };
+    if (dupeIndex > -1) {
+      savedContacts.contacts[dupeIndex] = newSavedContact;
+    } else {
+      savedContacts.contacts.push(newSavedContact);
+    }
+
+    // and save!
+    localStorage.setItem(CONTACTS_STORAGE_KEY, JSON.stringify(savedContacts));
+
+    return dupeIndex === -1;
   }
 
-  public addServer(code: AccessCode): Promise<Server> {
-    let token = parseInviteUrl(code);
-    if (!token) {
-      return Promise.reject(`Failed to parse access code: ${code}`);
+  public addServer(inviteUrl:string): Promise<void> {
```

Comment:
We've been using **Access Code** in the product specification. I think we should stick to that and move away from "invite" to avoid divergence between code and product.